### PR TITLE
vtgate: support stream grammar for messaging

### PR DESCRIPTION
**Pull request by derekperkins**
This adds AST support for Vitess specific grammar to allow for use with the Messaging component via vtgate.

### CODE REVIEW

File: go/vt/sqlparser/ast.go [hunk 1]
```diff
@@ -226,6 +226,11 @@ type Select struct {
 	Lock        string
 }
 
+// Stream represents a SELECT statement.
+type Stream struct {
+	Select
```
Comment (sougou) [inline go/vt/sqlparser/ast.go]: It's better to list the members here. That way, we can be more explicit about the constructs we don't support (like where clauses, etc).

In the future, if this becomes as complete as `Select`, we may be better off dropping this type and just adding a flag to `Select` indicating it's a stream.

File: go/vt/sqlparser/sql.y [hunk 2]
```diff
@@ -319,6 +320,12 @@ select_statement:
     $$ = &Select{Comments: Comments($2), Cache: $3, SelectExprs: SelectExprs{Nextval{Expr: $5}}, From: TableExprs{&AliasedTableExpr{Expr: $7}}}
   }
 
+stream_statement:
+  STREAM comment_opt cache_opt distinct_opt straight_join_opt select_expression_list from_opt where_expression_opt group_by_opt having_opt
```
Comment (sougou) [inline go/vt/sqlparser/sql.y]: Along the lines of the previous comment, let's simplify the syntax to allow only `STREAM comment_opt select_expression from_opt`. Note the singular select_expression.

File: go/vt/sqlparser/ast.go (hunk 1 above)
Comment (derekperkins) [reply go/vt/sqlparser/ast.go]: I started with adding an `action` field to `Select` like `Insert/Replace`, but didn't want to cause any performance regressions. I was more interested in inheriting the `Select` methods, but I can go ahead and copy them and reduce the functionality.

File: go/vt/sqlparser/sql.y [hunk 3]
```diff
@@ -319,6 +320,12 @@ select_statement:
     $$ = &Select{Comments: Comments($2), Cache: $3, SelectExprs: SelectExprs{Nextval{Expr: $5}}, From: TableExprs{&AliasedTableExpr{Expr: $7}}}
   }
 
+stream_statement:
+  STREAM comment_opt select_expression FROM table_name
+  {
+    $$ = &Stream{Comments: Comments($2), SelectExpr: $3, Table: $4}
```
Comment (yuananf) [inline go/vt/sqlparser/sql.y]: 'Table: $4' is wrong, should be 'Table: $5'

Comment (derekperkins) [reply go/vt/sqlparser/sql.y]: Thanks, that fixed it immediately!

File: go/vt/sqlparser/parse_test.go [hunk 4]
```diff
@@ -1089,6 +1089,8 @@ var (
 		input: "replace into t partition (p0) values (1, 'asdf')",
 	}, {
 		input: "delete from t partition (p0) where a = 1",
+	}, {
+		input: "stream * from t",
```
Comment (sougou) [inline go/vt/sqlparser/parse_test.go]: Need one extra construct: `stream /* comment */ * from t`

File: go/vt/vtgate/executor.go [hunk 5]
```diff
@@ -818,6 +825,49 @@ func (e *Executor) StreamExecute(ctx context.Context, method string, safeSession
...
+	stmt, err := sqlparser.Parse(sql)
+	if err != nil {
+		logStats.Error = err
+		return err
+	}
+
+	streamStmt, ok := stmt.(*sqlparser.Stream)
+	if !ok {
+		logStats.Error = err
+		return errors.New("preview identified a non-stream query as a stream query")
```
Comment (sougou) [inline go/vt/vtgate/executor.go]: This should be an internal error like here: https://github.com/youtube/vitess/blob/master/go/vt/vtgate/executor.go#L517.

File: go/vt/vtgate/executor_stream_test.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+	"github.com/youtube/vitess/go/sqltypes"
+	querypb "github.com/youtube/vitess/go/vt/proto/query"
+	topodatapb "github.com/youtube/vitess/go/vt/proto/topodata"
+	_ "github.com/youtube/vitess/go/vt/vtgate/vindexes"
+	"github.com/youtube/vitess/go/vt/vttablet/sandboxconn"
+	"golang.org/x/net/context"
+)
+
+func TestStreamSQLUnsharded(t *testing.T) {
```
Comment (sougou) [inline go/vt/vtgate/executor_stream_test.go]: Can you write one for Sharded?

File: go/vt/sqlparser/parse_test.go (hunk 4 above)
Comment (derekperkins) [reply go/vt/sqlparser/parse_test.go]: Done

File: go/vt/vtgate/executor.go (hunk 5 above)
Comment (derekperkins) [reply go/vt/vtgate/executor.go]: Good catch

File: go/vt/vtgate/executor_stream_test.go (hunk 6 above)
Comment (derekperkins) [reply go/vt/vtgate/executor_stream_test.go]: Yep

**Comment by derekperkins** [comment]
@sougou All your comments have been addressed, so this is ready for another review.

**Comment by sougou** [comment]
LGTM

**Comment by derekperkins** [comment]
Can you still not merge? :)

**Comment by sougou** [comment]
Nope :)

**Comment by derekperkins** [comment]
@alainjobart @enisoc I'd appreciate if you could merge this when you get a chance.

**Pull request merged**